Add timeout option to provider to wait for API

stbenjam opened the pull request:
If a timeout is specified, the provider will retry connecting to the API
until the timeout is reached. This can be useful when, for example,
another terraform provider is responsible for creating the Ironic
infrastructure. This is a hack, because terraform doesn't truly
understand depends_on relationships between providers.

This approach has limitations: it only works correctly on the first time
the infrastructure is created. If Ironic goes down you cannot rely on a
subsequent run of the other provider to repair Ironic, as Terraform will
expect to be able to talk to the API during the planning stage. This
isn't relevant to the Metal3 use case, as we only ever run apply once.

hardys commented on ironic/provider.go:
```diff
@@ -90,8 +132,52 @@ func configureProvider(schema *schema.ResourceData) (interface{}, error) {
+	defer close(timeout)
+
+	go func() {
+		for tries := 0; ; tries++ {
```
Would this code be simpler and more readable if we just calculated the maximum number of tries for a given timeout?

Obviously we'd need to ensuire the built in httpclient instance also has a sensible timeout to handle the case where the Get hangs up.

stbenjam replied:
The time.After built-in channel seems to be how a lot of people do timeouts in Go: https://gobyexample.com/timeouts

But, I could do it that way.  I don't think I could measure the length of the GET request itself, so the timeout could be off by a bit.

hardys replied:
Fair enough - I was thinking we'd set the GET timeout to a known value but if this is the common pattern then I'm not blocking on it :)

stbenjam replied:
Ok, I'd prefer to keep it as is as my bandwidth to take on more things is limited trying to shepherd the openshift/installer PR through. If the goroutine/`time.After` thing becomes an issue I can refactor it to do it the other way.

stbenjam commented:
I've updated this and added a basic test, I think it works but I noticed something odd locally with the goroutine timeout mechanism not working right in one case, will need to investigate and maybe refactor to do it in a less gopheristic way after all.

I'll try to test tomorrow with the installer and see how that goes.

hardys commented:
I tried this, and it appears to work without any timeout, but adding one results in this error:

```
level=debug msg="2019/07/11 15:30:32 [WARN] Provider \"libvirt\" produced an invalid plan for module.bootstrap.libvirt_ignition.bootstrap, but we are tolerating it because it is using the legacy plugin SDK."
level=debug msg="    The following problems may be the cause of any confusing errors from downstream operations:"
level=debug msg="      - .pool: planned value cty.StringVal(\"default\") does not match config value cty.NullVal(cty.String)"
level=debug msg="module.bootstrap.libvirt_ignition.bootstrap: Creating..."
level=debug msg="2019/07/11 15:30:32 [DEBUG] module.bootstrap.libvirt_ignition.bootstrap: applying the planned Create change"
level=debug msg="2019-07-11T15:30:32.935+0100 [DEBUG] plugin.terraform-provider-libvirt: 2019/07/11 15:30:32 [DEBUG] creating ignition file"
level=debug msg="2019/07/11 15:30:35 [DEBUG] module.masters.ironic_node_v1.openshift-master-host[0]: apply errored, but we're indicating that via the Error pointer rather than returning it: rpc error: code = Unavailable desc = transport is closing"
level=debug msg="2019/07/11 15:30:35 [ERROR] module.masters: eval: *terraform.EvalApplyPost, err: rpc error: code = Unavailable desc = transport is closing"
level=debug msg="2019/07/11 15:30:35 [ERROR] module.masters: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing"
level=debug msg="2019/07/11 15:30:35 [DEBUG] module.masters.ironic_node_v1.openshift-master-host[1]: apply errored, but we're indicating that via the Error pointer rather than returning it: rpc error: code = Unavailable desc = transport is closing"
level=debug msg="2019/07/11 15:30:35 [ERROR] module.masters: eval: *terraform.EvalApplyPost, err: rpc error: code = Unavailable desc = transport is closing"
level=debug msg="2019/07/11 15:30:35 [ERROR] module.masters: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing"
level=debug msg="2019/07/11 15:30:35 [DEBUG] module.masters.ironic_node_v1.openshift-master-host[2]: apply errored, but we're indicating that via the Error pointer rather than returning it: rpc error: code = Unavailable desc = transport is closing"
level=debug msg="2019/07/11 15:30:35 [ERROR] module.masters: eval: *terraform.EvalApplyPost, err: rpc error: code = Unavailable desc = transport is closing"
level=debug msg="2019/07/11 15:30:35 [ERROR] module.masters: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing"

```

stbenjam commented:
> I tried this, and it appears to work without any timeout, but adding one results in this error:
> 
> ```
> level=debug msg="2019/07/11 15:30:32 [WARN] Provider \"libvirt\" produced an invalid plan for module.bootstrap.libvirt_ignition.bootstrap, but we are tolerating it because it is using the legacy plugin SDK."
> level=debug msg="    The following problems may be the cause of any confusing errors from downstream operations:"
> level=debug msg="      - .pool: planned value cty.StringVal(\"default\") does not match config value cty.NullVal(cty.String)"
> level=debug msg="module.bootstrap.libvirt_ignition.bootstrap: Creating..."
> level=debug msg="2019/07/11 15:30:32 [DEBUG] module.bootstrap.libvirt_ignition.bootstrap: applying the planned Create change"
> level=debug msg="2019-07-11T15:30:32.935+0100 [DEBUG] plugin.terraform-provider-libvirt: 2019/07/11 15:30:32 [DEBUG] creating ignition file"
> level=debug msg="2019/07/11 15:30:35 [DEBUG] module.masters.ironic_node_v1.openshift-master-host[0]: apply errored, but we're indicating that via the Error pointer rather than returning it: rpc error: code = Unavailable desc = transport is closing"
> level=debug msg="2019/07/11 15:30:35 [ERROR] module.masters: eval: *terraform.EvalApplyPost, err: rpc error: code = Unavailable desc = transport is closing"
> level=debug msg="2019/07/11 15:30:35 [ERROR] module.masters: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing"
> level=debug msg="2019/07/11 15:30:35 [DEBUG] module.masters.ironic_node_v1.openshift-master-host[1]: apply errored, but we're indicating that via the Error pointer rather than returning it: rpc error: code = Unavailable desc = transport is closing"
> level=debug msg="2019/07/11 15:30:35 [ERROR] module.masters: eval: *terraform.EvalApplyPost, err: rpc error: code = Unavailable desc = transport is closing"
> level=debug msg="2019/07/11 15:30:35 [ERROR] module.masters: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing"
> level=debug msg="2019/07/11 15:30:35 [DEBUG] module.masters.ironic_node_v1.openshift-master-host[2]: apply errored, but we're indicating that via the Error pointer rather than returning it: rpc error: code = Unavailable desc = transport is closing"
> level=debug msg="2019/07/11 15:30:35 [ERROR] module.masters: eval: *terraform.EvalApplyPost, err: rpc error: code = Unavailable desc = transport is closing"
> level=debug msg="2019/07/11 15:30:35 [ERROR] module.masters: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing"
> ```

This one was just a small mistake on my part, I was looking at the http response without checking error. That's fixed, but now running into another odd issue where my time.Sleep never wakes up again, even after rewriting this with the less clever approach.

stbenjam commented on ironic/provider.go:
```diff
@@ -90,8 +134,42 @@ func configureProvider(schema *schema.ResourceData) (interface{}, error) {
+			}
+		}
+
+		time.Sleep(5 * time.Second)
```
This function is not clever at all anymore, but we are still hitting a really weird problem on this sleep call.  If you call `time.Sleep` here, the Go scheduler never wakes up this goroutine again. It stays stuck forever.  The tests work fine, this only happens when running inside terraform in the installer.  There's a lot of goroutines (about 25) when this runs in the installer, but my box has GOMAXPROCS to 48, so there's not any resource starvation going on AFAICT.

I spent some time on the gopher slack and talked to some people who gave me suggestions about some debugging tools, to see what's going on (https://jvns.ca/blog/2017/09/24/profiling-go-with-pprof/).  I didn't see anything obvious other than seeing that the goroutine never wakes up again, it shows [sleep] next to it forever.

I'll keep on this next week, I'm on PTO tomorrow. I'm going to simplify the problem down to a bare bones terraform provider and see if any time.Sleep in terraform causes it to lock up, and then go from there.

The only workaround I can think of at the moment is a busy for loop instead of sleeping...but that's a terrible idea, TBH.

stbenjam replied:
It was not a concurrency issue.  time.Sleep was working, but after the first log gets printed, and the goroutine goes to sleep for n seconds, something is changing the log output to go to `ioutil.Discard`.  There's some suspicious looking code in Terraform, going to investigate more next week.

stbenjam commented:
@hardys I think I've fixed the issues, give this another try.  Note the comment here: https://github.com/openshift-metal3/terraform-provider-ironic/pull/30/files#diff-095d41bfe93710c481ac45e43e71cc0eR199

You will not see output for tries after the first, but it *is* trying, and then should proceed when Ironic API is up.

hardys commented:
Thanks @stbenjam this is working for me now - the only issue we may face is when you download the RHCOS image on the bootstrap node we're going to be quite close to the 30min timeout unless it's cached.

The CI shows pending but clicking Details shows it ran OK, so I'll go ahead and merge.
